### PR TITLE
SkyDNS is the only NS for Pods with DNSPolicy=ClusterFirst

### PR DESCRIPTION
**Pull request by ArtfulCoder**
Addresses issue : https://github.com/kubernetes/kubernetes/issues/15592
See https://github.com/kubernetes/kubernetes/issues/15592#issue-111306439

### CODE REVIEW

**Comment by vishh** [comment]
@ArtfulCoder: Can you add a brief description on why this change is required?

**Comment by ArtfulCoder** [comment]
Sure . I can copy it from on the issue I have linked already.

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1190,7 +1190,7 @@ func (kl *Kubelet) getClusterDNS(pod *api.Pod) ([]string, []string, error) {
 	var dns, dnsSearch []string
 
 	if kl.clusterDNS != nil {
-		dns = append([]string{kl.clusterDNS.String()}, hostDNS...)
+		dns = []string{kl.clusterDNS.String()}
```
Comment (thockin) [inline pkg/kubelet/kubelet.go]: Comment on this, please.  It's less obvious now.  Also comment on L1163 is now wrong.  Please audit this function to make sure comments match reality

**Comment by ArtfulCoder** [comment]
Punting this PR to 1.2 (after https://github.com/kubernetes/kubernetes/issues/15752 is addressed)

**Comment by ArtfulCoder** [comment]
@thockin PTAL

**Comment by ArtfulCoder** [comment]
@dchen1107 since its a kubelet change

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -1330,9 +1330,17 @@ func (kl *Kubelet) getClusterDNS(pod *api.Pod) ([]string, []string, error) {
 	var dns, dnsSearch []string
 
 	if kl.clusterDNS != nil {
-		dns = append([]string{kl.clusterDNS.String()}, hostDNS...)
+		// for a pod with DnsClusterFirst policy, the cluster DNS server is the only nameserver configured for
```
Comment (thockin) [inline pkg/kubelet/kubelet.go]: nit: DNSClusterFirst

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -1330,9 +1330,17 @@ func (kl *Kubelet) getClusterDNS(pod *api.Pod) ([]string, []string, error) {
...
+		// the pod. The cluster DNS server itself will forward queries to other nameservers that is configured to use,
+		// in case the cluster DNS server cannot resolve the DNS query itself
+		dns = []string{kl.clusterDNS.String()}
 	} else {
-		dns = hostDNS
+		// clusterDNS is not known.
+		// pod with ClusterDNSFirstPolicy cannot be created
+		log := fmt.Sprintf("kubelet on %q does not have ClusterDNS IP configured and cannot create Pod using ClusterDNSFirst policy. pod:%q", kl.nodeName, kubecontainer.GetPodFullName(pod))
```
Comment (thockin) [inline pkg/kubelet/kubelet.go]: s/ClusterDNSFirst/%s and expand the api.DNSClusterFirst value

**Comment by thockin** [comment]
nits then LGTM

**Comment by thockin** [comment]
status on this?

**Comment by ArtfulCoder** [comment]
will incorporate your feedback (was waiting for  iptables proxy to be default)

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (ArtfulCoder) [inline pkg/kubelet/kubelet.go]: done

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (ArtfulCoder) [inline pkg/kubelet/kubelet.go]: done

**Comment by ArtfulCoder** [comment]
@thockin PTAL

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (thockin) [inline pkg/kubelet/kubelet.go]: isn't the node name and component already in the event structure?

**Comment by thockin** [comment]
LGTM - just audit the event for duplicated info between fields and string

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (ArtfulCoder) [inline pkg/kubelet/kubelet.go]: done

**Comment by ArtfulCoder** [comment]
@k8s-bot test this please

**Comment by eparis** [comment]
did you intend to include those systrace files?

**Comment by ArtfulCoder** [comment]
I didn't. I fixed it :)

**Comment by ArtfulCoder** [comment]
e2e failure is unrelated.
http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-build-test-e2e-gce/18006/

**Comment by ArtfulCoder** [comment]
@k8s-bot test this please

**Comment by ArtfulCoder** [comment]
Incorporated Tim's feedback and re-added LGTM label that got removed because I update the PR with the feeback.

**Comment by ArtfulCoder** [comment]
@k8s-bot test this

**Comment by ArtfulCoder** [comment]
@k8s-bot test this

**Comment by wojtek-t** [comment]
@thockin @ArtfulCoder @gmarek 

Just to give more clear context why we decided to revert it.
- it broke kubemark-gce, which is currently a blocking build for mergebot

A bit more context.
- in kubemark, there is no skydns (we want to add it at some point, but that's not done yet)
- however, with this PR merge, this caused that no pods were started (all the pods were successfully created & scheduled, but not even a single pod was started)
- note, that we are using "pause" images, which aren't the most demanding ones :)

 So I think the question is the following:
  if this change requires DNS to start __any__ pod, it has be clearly stated (I'm not saying it's not reasonable - it makes sense to say that). 
  However, if it wasn't the intention of this PR, then there was some bug in it.

**Comment by gmarek** [comment]
If we're going to require to DNS to start any pod, we probably should remove an option to disable it through config.

**Comment by eparis** [comment]
I'll throw in my objection to 'dns is required' until dns is a part of the system, instead of something on top of the system. aka I think @DirectXMan12  did some work to embed dns inside each kubelet automatically, instead of having to do something magic after the cluster is up. (not sure where it ended up). So i'm against such a new requirement as things stand today.

Real world example, the kubernetes cluster where the submit queue bot runs doesn't have dns and seems to work just fine.

**Comment by ArtfulCoder** [comment]
We don't need to have DNS in the cluster.
This also means that, we should have pods with DNSpolicy of ClusterFirst.
If we change the pod spec in the test to not use CluaterFirst DNSPolicy, the test will be green again.

I feel this is correct. We should not allow such pods to run when there is no cluster DNS server.

**Comment by wojtek-t** [comment]
@ArtfulCoder - got it thanks (I think I just didn't understand it).

@gmarek - can you please fix density test (that kubemark is running) so that all pods there  use different dns policy. And then we should re-revert this PR.

**Comment by gmarek** [comment]
The thing is that we were not explicitly setting any DNSpolicy, and this change broke us. Can we make DNSDefault a default value?

**Comment by ArtfulCoder** [comment]
Tim suggested that we just fall back to DNSDefault if dns server is not
available and log a warning.
I will make the change and it should fix these issues.

**Comment by gmarek** [comment]
Perfect - thanks. In case it is needed: #18081

**Pull request merged**